app/src/lib.rs

```diff
@@ -30,15 +30,42 @@ pub struct CommandRequest<C: DeserializeOwned + Serialize> {
     pub command: Option<C>,
 }
 
-/// An EventRequest may only be emitted by a server, of which there can be many, and
-/// only in relation to having received a [CommandRequest] from a client. Event requests
-/// take a type that provides their identifier; usually an enum. Event requests convey
+/// A temporal event is one that has its durability conveyed.
+pub trait TemporalEvent: Clone + DeserializeOwned + Serialize {}
+
+/// An event that has been logged, providing their identifier; usually an enum. These replies convey
 /// the offset they are associated with. If an offset overflows to zero then it is the
 /// server's responsibility to convey any important events that the client may need.
 /// It is the client's responsibility to clear state in relation to previous events when
 /// an offset less than or equal to the one it requested. An event request also conveys a
 /// delta in time in a form that the client and its servers understand, and relative to
 /// the server's current notion of time.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Logged<E>(pub E, pub u32);
+impl<E: Clone + DeserializeOwned + Serialize> TemporalEvent for Logged<E> {}
+
+/// An event that has not been logged by the server and may be consumed by the client,
+/// often to convey some instantaneous event that does not need to be recorded. Events
+/// of this category should be benign if they are not consumed by a client.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+pub struct Ephemeral<E>(pub E);
+impl<E: Clone + DeserializeOwned + Serialize> TemporalEvent for Ephemeral<E> {}
+
+/// Either a logged or an ephemeral event can be declared for use.
+#[derive(Clone, Debug, Deserialize, Eq, PartialEq, Serialize)]
+#[non_exhaustive]
+pub enum Either<E, EE> {
```

Comment:
`Either` is an interesting choice of name here - it's very general in the realm of programming, especially functional programming. I wonder if there's a more domain-specific name. Is `Event` overused?

Reply:
I chose `Either` as it directly conveyed what it is here and looks good when used... More directly it'd be `EitherLoggedOrEphemeral`, but that's too long. `Either` also often turns up in domain-specific places, like Embassy: https://github.com/embassy-rs/embassy/blob/master/embassy-futures/src/select.rs#L10 and also futures-rs I think.

Reply:
Ok, fair enough. Interesting that it appears in embassy too.

Reply:
Great